feat(move): fully implement Dragon Darts

Aether21 opened the pull request:
<!-- (Once you have read these comments, you are free to remove them) -->
<!-- Feel free to look at other PRs for examples -->
<!--
Make sure the title includes categorization (choose the one that best fits):
-  [Bug]: If the PR is primarily a bug fix
-  [Move]: If a move has new or changed functionality
-  [Ability]: If an ability has new or changed functionality
-  [Item]: For new or modified items
-  [Mystery]: For new or modified Mystery Encounters
-  [Test]: If the PR is primarily adding or modifying tests
-  [UI/UX]: If the PR is changing UI/UX elements
-  [Audio]: If the PR is adding or changing music/sfx (may come with an associated assets PR)
-  [Sprite]: If the PR is adding or changing sprites (may come with an associated assets PR)
-  [Balance]: If the PR is related to game balance
-  [Challenge]: If the PR is adding or modifying challenges
-  [Refactor]: If the PR is primarily rewriting existing code
-  [Dev]: If the PR is primarily changing something pertaining to development (lefthook hooks, linter rules, etc.)
-  [i18n]: If the PR is primarily adding/changing locale keys or key usage (may come with an associated locales PR)
-  [Docs]: If the PR is adding or modifying documentation (such as tsdocs/code comments)
-  [GitHub]: For changes to GitHub workflows/templates/etc
-  [Misc]: If no other category fits the PR
-->

## Credit
This PR would not be possible without the [hard work of innerthunder on the Poketernity project](https://github.com/Despair-Games/poketernity/pull/46)! Thank you very much for letting me translate this code to PokeRogue!

## What are the changes the user will see?
Dragon Darts now has "smart" targeting, thereby fully implementing the move (save for a few edge cases)

## Why am I making these changes?
Dragon Darts deserves to be fully realized. Also I wanted to use it for mono Dragon runs.

## What are the changes from a developer perspective?
<!--
Explicitly state what are the changes introduced by the PR
You can make use of a comparison between what was the state before and after your PR changes
Ex: What files have been changed? What classes/functions/variables/etc have been added or changed?
-->
- `enums/move-target`: New MoveTarget option `DRAGON_DARTS` to implement Dragon Darts' dynamic targeting.
- `phases/move-effect-phase:`
    - `DRAGON_DARTS` move targeting adjusts the move's target for the given strike in double battles in two ways:
        - When a new strike for the move is queued, the target for the next hit is set to the prior target's ally (if on field)
        - If the strike would be predicted to fail its hit check against the base target, the move shifts targets to the base target's ally before applying effects.
    - `hitCheck` and `protectedCheck` now take a simulated argument to stop Dragon Darts from activating effects during its redirection. The logic to apply protection effects has been slightly edited to better support this change.
- `test/moves/dragon_darts.test:` Test suite on Dragon Darts' smart targeting based on innerthunder's PR (with enhancements to use new testing utils/add details)

## Screenshots/Videos
<!--
If your changes are changing anything on the user experience, please provide visual proofs of it
Please take screenshots/videos before and after your changes, to show what is brought by this PR
-->

## How to test the changes?
pnpm run test dragon_darts

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`pnpm test:silent`)
  - [x] Have I created new automated tests (`pnpm test:create`) or updated existing tests related to the PR's changes?
- [x] Have I provided screenshots/videos of the changes (if applicable)?
  - [x] Have I made sure that any UI change works for both UI themes (default and legacy)?

Are there any localization additions or changes? If so:
- [n/a] Has a locales PR been created on the [locales](https://github.com/pagefaultgames/pokerogue-locales) repo?
  - [n/a] If so, please leave a link to it here: 
- [n/a] Has the translation team been contacted for proofreading/translation?

Does this require any changes to the assets folder? If so:
  - [n/a] Has a PR been created on the [assets](https://github.com/pagefaultgames/pokerogue-assets) repo?
    - [n/a] If so, please leave a link to it here:

Aether21 committed:
initial skeleton for smart targeting

Aether21 committed:
initial state of smart targeting for dragon darts

Aether21 committed:
add edge case to dragon darts move.ts

Aether21 committed:
add test file for dragon darts

Aether21 committed:
switching final MoveResult case for Center of Attention test

Aether21 committed:
editing onContact protect logic to handle dragon darts redirect

Aether21 committed:
add dragon darts to switch moveTarget case

Aether21 committed:
finished smart targeting logic, including adding simulated checks to hitCheck and protectedCheck

Bertie690 commented on src/battle-scene.ts:
```diff
@@ -825,6 +826,15 @@ export class BattleScene extends SceneBase {
+  public getPokemonByBattlerIndex(battlerIndex: BattlerIndex): Pokemon | undefined {
+    return this.getField().find(p => p.getBattlerIndex() === battlerIndex);
+  }
+
```
This is equivalent to `this.getField()[battlerIndex]` for really weird and fucked up reasons.

In hindsight, a helper function like this would be a decent abstraction over how exactly `BattlerIndex`es work (or don't), so I might steal this and implement it across the codebase. :)

Bertie690 commented on src/data/moves/move.ts:
I'd prefer to write out the edge cases manually for documentation purposes

fabske0 commented on src/data/battler-tags.ts:
```diff
@@ -1767,8 +1782,10 @@ export class ProtectedTag extends BattlerTag {
-  lapse(pokemon: Pokemon, lapseType: BattlerTagLapseType): boolean {
-    if (lapseType === BattlerTagLapseType.CUSTOM) {
+  // lapse(pokemon: Pokemon, lapseType: BattlerTagLapseType): boolean {
+  //   if (lapseType === BattlerTagLapseType.CUSTOM) {
```
```suggestion
```

fabske0 commented on src/data/battler-tags.ts:
```diff
@@ -1784,7 +1801,8 @@ export class ProtectedTag extends BattlerTag {
     }
 
-    return super.lapse(pokemon, lapseType);
+    //return super.lapse(pokemon, lapseType);
```
```suggestion
```

fabske0 reviewed (commented):
I guess these can be removed?

Aether21 committed:
remove unnecessary code in src/data/battler-tags.ts

Co-authored-by: Fabi <192151969+fabske0@users.noreply.github.com>

Aether21 committed:
remove unnecessary code in src/data/battler-tags.ts

Co-authored-by: Fabi <192151969+fabske0@users.noreply.github.com>

Aether21 committed:
Update src/data/battler-tags.ts

Co-authored-by: Fabi <192151969+fabske0@users.noreply.github.com>

Aether21 committed:
Update src/data/battler-tags.ts

Co-authored-by: Fabi <192151969+fabske0@users.noreply.github.com>

Aether21 committed:
Update src/data/moves/move.ts

Co-authored-by: Fabi <192151969+fabske0@users.noreply.github.com>

Aether21 committed:
add double protect edge case to move.ts

Bertie690 committed:
Assorted minor cleanups for code review

# Conflicts:
#	src/data/battler-tags.ts

Bertie690 committed:
fixed test file up a bit

DayKev committed:
Merge branch 'beta' into dragon-darts-fix

DayKev commented on src/data/moves/move.ts:
```diff
@@ -11689,7 +11700,10 @@ export function initMoves() {
+      .target(MoveTarget.DRAGON_DARTS)
+      .edgeCase(), 
+      // see `dragon_darts.test.ts` for documented edge cases
+      // when both enemies use protect, both hits go into the ally of the current target
```
```suggestion
      // see `dragon_darts.test.ts` for documented edge cases
      // when both enemies use protect, both hits go into the ally of the current target
      .edgeCase(),
```

DayKev commented on src/phases/move-effect-phase.ts:
```diff
@@ -355,6 +386,21 @@ export class MoveEffectPhase extends PokemonPhase {
       return;
     }
 
+    /**
```
```suggestion
    /*
```

Aether21 committed:
Update src/data/moves/move.ts

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

Aether21 committed:
Update src/phases/move-effect-phase.ts

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

Aether21 committed:
Update src/phases/move-effect-phase.ts

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

Aether21 committed:
Update src/phases/move-effect-phase.ts

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

DayKev committed:
Apply Biome

DayKev committed:
Merge branch 'beta' into dragon-darts-fix

DayKev committed:
Make variable names correct

This is why `enemy1`/etc is better than `pokemonname1`/etc,
it won't be out of sync with the actual pokemon

Bertie690 committed:
Hopefully fixed Ddarts bug + general housekeeping

Bertie690 committed:
Moved around code, fixed most remaining bugs

Aether21 committed:
fixed both enemy immune, trigger effects once test

Aether21 committed:
enabled passing should not trigger ability effects test

Aether21 committed:
fixing redirect to ally if missed original target tests

fabske0 committed:
run biome

Aether21 committed:
updated ContactProtectedTag docs

DayKev committed:
Merge branch 'beta' into dragon-darts-fix
